// @zenstackhq/language/generated/ast.js
"use strict";
/******************************************************************************
 * This file was generated by langium-cli 1.3.1.
 * DO NOT EDIT MANUALLY!
 ******************************************************************************/
Object.defineProperty(exports, "__esModule", { value: true });
exports.TypeDefField = exports.TypeDef = exports.ThisExpr = exports.StringLiteral = exports.ReferenceExpr = exports.ReferenceArg = exports.PluginField = exports.Plugin = exports.ObjectExpr = exports.NumberLiteral = exports.NullExpr = exports.ModelImport = exports.Model = exports.MemberAccessExpr = exports.InvocationExpr = exports.InternalAttribute = exports.GeneratorDecl = exports.FunctionParamType = exports.FunctionParam = exports.FunctionDecl = exports.FieldInitializer = exports.EnumField = exports.Enum = exports.DataSource = exports.DataModelFieldType = exports.DataModelFieldAttribute = exports.DataModelField = exports.DataModelAttribute = exports.DataModel = exports.ConfigInvocationExpr = exports.ConfigInvocationArg = exports.ConfigField = exports.ConfigArrayExpr = exports.BooleanLiteral = exports.BinaryExpr = exports.AttributeParamType = exports.AttributeParam = exports.AttributeArg = exports.Attribute = exports.ArrayExpr = exports.Argument = exports.TypeDefFieldTypes = exports.TypeDeclaration = exports.ReferenceTarget = exports.MemberAccessTarget = exports.LiteralExpr = exports.Expression = exports.ConfigExpr = exports.AbstractDeclaration = exports.ZModelTerminals = void 0;
exports.reflection = exports.ZModelAstReflection = exports.UnsupportedFieldType = exports.UnaryExpr = exports.TypeDefFieldType = void 0;
exports.isAbstractDeclaration = isAbstractDeclaration;
exports.isBoolean = isBoolean;
exports.isBuiltinType = isBuiltinType;
exports.isConfigExpr = isConfigExpr;
exports.isExpression = isExpression;
exports.isExpressionType = isExpressionType;
exports.isLiteralExpr = isLiteralExpr;
exports.isMemberAccessTarget = isMemberAccessTarget;
exports.isReferenceTarget = isReferenceTarget;
exports.isRegularID = isRegularID;
exports.isRegularIDWithTypeNames = isRegularIDWithTypeNames;
exports.isTypeDeclaration = isTypeDeclaration;
exports.isTypeDefFieldTypes = isTypeDefFieldTypes;
exports.isArgument = isArgument;
exports.isArrayExpr = isArrayExpr;
exports.isAttribute = isAttribute;
exports.isAttributeArg = isAttributeArg;
exports.isAttributeParam = isAttributeParam;
exports.isAttributeParamType = isAttributeParamType;
exports.isBinaryExpr = isBinaryExpr;
exports.isBooleanLiteral = isBooleanLiteral;
exports.isConfigArrayExpr = isConfigArrayExpr;
exports.isConfigField = isConfigField;
exports.isConfigInvocationArg = isConfigInvocationArg;
exports.isConfigInvocationExpr = isConfigInvocationExpr;
exports.isDataModel = isDataModel;
exports.isDataModelAttribute = isDataModelAttribute;
exports.isDataModelField = isDataModelField;
exports.isDataModelFieldAttribute = isDataModelFieldAttribute;
exports.isDataModelFieldType = isDataModelFieldType;
exports.isDataSource = isDataSource;
exports.isEnum = isEnum;
exports.isEnumField = isEnumField;
exports.isFieldInitializer = isFieldInitializer;
exports.isFunctionDecl = isFunctionDecl;
exports.isFunctionParam = isFunctionParam;
exports.isFunctionParamType = isFunctionParamType;
exports.isGeneratorDecl = isGeneratorDecl;
exports.isInternalAttribute = isInternalAttribute;
exports.isInvocationExpr = isInvocationExpr;
exports.isMemberAccessExpr = isMemberAccessExpr;
exports.isModel = isModel;
exports.isModelImport = isModelImport;
exports.isNullExpr = isNullExpr;
exports.isNumberLiteral = isNumberLiteral;
exports.isObjectExpr = isObjectExpr;
exports.isPlugin = isPlugin;
exports.isPluginField = isPluginField;
exports.isReferenceArg = isReferenceArg;
exports.isReferenceExpr = isReferenceExpr;
exports.isStringLiteral = isStringLiteral;
exports.isThisExpr = isThisExpr;
exports.isTypeDef = isTypeDef;
exports.isTypeDefField = isTypeDefField;
exports.isTypeDefFieldType = isTypeDefFieldType;
exports.isUnaryExpr = isUnaryExpr;
exports.isUnsupportedFieldType = isUnsupportedFieldType;
const langium_1 = require("langium");
exports.ZModelTerminals = {
    WS: /\s+/,
    INTERNAL_ATTRIBUTE_NAME: /@@@([_a-zA-Z][\w_]*\.)*[_a-zA-Z][\w_]*/,
    MODEL_ATTRIBUTE_NAME: /@@([_a-zA-Z][\w_]*\.)*[_a-zA-Z][\w_]*/,
    FIELD_ATTRIBUTE_NAME: /@([_a-zA-Z][\w_]*\.)*[_a-zA-Z][\w_]*/,
    ID: /[_a-zA-Z][\w_]*/,
    STRING: /"(\\.|[^"\\])*"|'(\\.|[^'\\])*'/,
    NUMBER: /[+-]?[0-9]+(\.[0-9]+)?/,
    TRIPLE_SLASH_COMMENT: /\/\/\/[^\n\r]*/,
    ML_COMMENT: /\/\*[\s\S]*?\*\//,
    SL_COMMENT: /\/\/[^\n\r]*/,
};
exports.AbstractDeclaration = 'AbstractDeclaration';
function isAbstractDeclaration(item) {
    return exports.reflection.isInstance(item, exports.AbstractDeclaration);
}
function isBoolean(item) {
    return typeof item === 'boolean';
}
function isBuiltinType(item) {
    return item === 'String' || item === 'Boolean' || item === 'Int' || item === 'BigInt' || item === 'Float' || item === 'Decimal' || item === 'DateTime' || item === 'Json' || item === 'Bytes';
}
exports.ConfigExpr = 'ConfigExpr';
function isConfigExpr(item) {
    return exports.reflection.isInstance(item, exports.ConfigExpr);
}
exports.Expression = 'Expression';
function isExpression(item) {
    return exports.reflection.isInstance(item, exports.Expression);
}
function isExpressionType(item) {
    return item === 'String' || item === 'Int' || item === 'Float' || item === 'Boolean' || item === 'DateTime' || item === 'Null' || item === 'Object' || item === 'Any' || item === 'Unsupported';
}
exports.LiteralExpr = 'LiteralExpr';
function isLiteralExpr(item) {
    return exports.reflection.isInstance(item, exports.LiteralExpr);
}
exports.MemberAccessTarget = 'MemberAccessTarget';
function isMemberAccessTarget(item) {
    return exports.reflection.isInstance(item, exports.MemberAccessTarget);
}
exports.ReferenceTarget = 'ReferenceTarget';
function isReferenceTarget(item) {
    return exports.reflection.isInstance(item, exports.ReferenceTarget);
}
function isRegularID(item) {
    return item === 'model' || item === 'enum' || item === 'attribute' || item === 'datasource' || item === 'plugin' || item === 'abstract' || item === 'in' || item === 'view' || item === 'import' || item === 'type' || (typeof item === 'string' && (/[_a-zA-Z][\w_]*/.test(item)));
}
function isRegularIDWithTypeNames(item) {
    return isRegularID(item) || item === 'String' || item === 'Boolean' || item === 'Int' || item === 'BigInt' || item === 'Float' || item === 'Decimal' || item === 'DateTime' || item === 'Json' || item === 'Bytes' || item === 'Null' || item === 'Object' || item === 'Any' || item === 'Unsupported';
}
exports.TypeDeclaration = 'TypeDeclaration';
function isTypeDeclaration(item) {
    return exports.reflection.isInstance(item, exports.TypeDeclaration);
}
exports.TypeDefFieldTypes = 'TypeDefFieldTypes';
function isTypeDefFieldTypes(item) {
    return exports.reflection.isInstance(item, exports.TypeDefFieldTypes);
}
exports.Argument = 'Argument';
function isArgument(item) {
    return exports.reflection.isInstance(item, exports.Argument);
}
exports.ArrayExpr = 'ArrayExpr';
function isArrayExpr(item) {
    return exports.reflection.isInstance(item, exports.ArrayExpr);
}
exports.Attribute = 'Attribute';
function isAttribute(item) {
    return exports.reflection.isInstance(item, exports.Attribute);
}
exports.AttributeArg = 'AttributeArg';
function isAttributeArg(item) {
    return exports.reflection.isInstance(item, exports.AttributeArg);
}
exports.AttributeParam = 'AttributeParam';
function isAttributeParam(item) {
    return exports.reflection.isInstance(item, exports.AttributeParam);
}
exports.AttributeParamType = 'AttributeParamType';
function isAttributeParamType(item) {
    return exports.reflection.isInstance(item, exports.AttributeParamType);
}
exports.BinaryExpr = 'BinaryExpr';
function isBinaryExpr(item) {
    return exports.reflection.isInstance(item, exports.BinaryExpr);
}
exports.BooleanLiteral = 'BooleanLiteral';
function isBooleanLiteral(item) {
    return exports.reflection.isInstance(item, exports.BooleanLiteral);
}
exports.ConfigArrayExpr = 'ConfigArrayExpr';
function isConfigArrayExpr(item) {
    return exports.reflection.isInstance(item, exports.ConfigArrayExpr);
}
exports.ConfigField = 'ConfigField';
function isConfigField(item) {
    return exports.reflection.isInstance(item, exports.ConfigField);
}
exports.ConfigInvocationArg = 'ConfigInvocationArg';
function isConfigInvocationArg(item) {
    return exports.reflection.isInstance(item, exports.ConfigInvocationArg);
}
exports.ConfigInvocationExpr = 'ConfigInvocationExpr';
function isConfigInvocationExpr(item) {
    return exports.reflection.isInstance(item, exports.ConfigInvocationExpr);
}
exports.DataModel = 'DataModel';
function isDataModel(item) {
    return exports.reflection.isInstance(item, exports.DataModel);
}
exports.DataModelAttribute = 'DataModelAttribute';
function isDataModelAttribute(item) {
    return exports.reflection.isInstance(item, exports.DataModelAttribute);
}
exports.DataModelField = 'DataModelField';
function isDataModelField(item) {
    return exports.reflection.isInstance(item, exports.DataModelField);
}
exports.DataModelFieldAttribute = 'DataModelFieldAttribute';
function isDataModelFieldAttribute(item) {
    return exports.reflection.isInstance(item, exports.DataModelFieldAttribute);
}
exports.DataModelFieldType = 'DataModelFieldType';
function isDataModelFieldType(item) {
    return exports.reflection.isInstance(item, exports.DataModelFieldType);
}
exports.DataSource = 'DataSource';
function isDataSource(item) {
    return exports.reflection.isInstance(item, exports.DataSource);
}
exports.Enum = 'Enum';
function isEnum(item) {
    return exports.reflection.isInstance(item, exports.Enum);
}
exports.EnumField = 'EnumField';
function isEnumField(item) {
    return exports.reflection.isInstance(item, exports.EnumField);
}
exports.FieldInitializer = 'FieldInitializer';
function isFieldInitializer(item) {
    return exports.reflection.isInstance(item, exports.FieldInitializer);
}
exports.FunctionDecl = 'FunctionDecl';
function isFunctionDecl(item) {
    return exports.reflection.isInstance(item, exports.FunctionDecl);
}
exports.FunctionParam = 'FunctionParam';
function isFunctionParam(item) {
    return exports.reflection.isInstance(item, exports.FunctionParam);
}
exports.FunctionParamType = 'FunctionParamType';
function isFunctionParamType(item) {
    return exports.reflection.isInstance(item, exports.FunctionParamType);
}
exports.GeneratorDecl = 'GeneratorDecl';
function isGeneratorDecl(item) {
    return exports.reflection.isInstance(item, exports.GeneratorDecl);
}
exports.InternalAttribute = 'InternalAttribute';
function isInternalAttribute(item) {
    return exports.reflection.isInstance(item, exports.InternalAttribute);
}
exports.InvocationExpr = 'InvocationExpr';
function isInvocationExpr(item) {
    return exports.reflection.isInstance(item, exports.InvocationExpr);
}
exports.MemberAccessExpr = 'MemberAccessExpr';
function isMemberAccessExpr(item) {
    return exports.reflection.isInstance(item, exports.MemberAccessExpr);
}
exports.Model = 'Model';
function isModel(item) {
    return exports.reflection.isInstance(item, exports.Model);
}
exports.ModelImport = 'ModelImport';
function isModelImport(item) {
    return exports.reflection.isInstance(item, exports.ModelImport);
}
exports.NullExpr = 'NullExpr';
function isNullExpr(item) {
    return exports.reflection.isInstance(item, exports.NullExpr);
}
exports.NumberLiteral = 'NumberLiteral';
function isNumberLiteral(item) {
    return exports.reflection.isInstance(item, exports.NumberLiteral);
}
exports.ObjectExpr = 'ObjectExpr';
function isObjectExpr(item) {
    return exports.reflection.isInstance(item, exports.ObjectExpr);
}
exports.Plugin = 'Plugin';
function isPlugin(item) {
    return exports.reflection.isInstance(item, exports.Plugin);
}
exports.PluginField = 'PluginField';
function isPluginField(item) {
    return exports.reflection.isInstance(item, exports.PluginField);
}
exports.ReferenceArg = 'ReferenceArg';
function isReferenceArg(item) {
    return exports.reflection.isInstance(item, exports.ReferenceArg);
}
exports.ReferenceExpr = 'ReferenceExpr';
function isReferenceExpr(item) {
    return exports.reflection.isInstance(item, exports.ReferenceExpr);
}
exports.StringLiteral = 'StringLiteral';
function isStringLiteral(item) {
    return exports.reflection.isInstance(item, exports.StringLiteral);
}
exports.ThisExpr = 'ThisExpr';
function isThisExpr(item) {
    return exports.reflection.isInstance(item, exports.ThisExpr);
}
exports.TypeDef = 'TypeDef';
function isTypeDef(item) {
    return exports.reflection.isInstance(item, exports.TypeDef);
}
exports.TypeDefField = 'TypeDefField';
function isTypeDefField(item) {
    return exports.reflection.isInstance(item, exports.TypeDefField);
}
exports.TypeDefFieldType = 'TypeDefFieldType';
function isTypeDefFieldType(item) {
    return exports.reflection.isInstance(item, exports.TypeDefFieldType);
}
exports.UnaryExpr = 'UnaryExpr';
function isUnaryExpr(item) {
    return exports.reflection.isInstance(item, exports.UnaryExpr);
}
exports.UnsupportedFieldType = 'UnsupportedFieldType';
function isUnsupportedFieldType(item) {
    return exports.reflection.isInstance(item, exports.UnsupportedFieldType);
}
class ZModelAstReflection extends langium_1.AbstractAstReflection {
    getAllTypes() {
        return ['AbstractDeclaration', 'Argument', 'ArrayExpr', 'Attribute', 'AttributeArg', 'AttributeParam', 'AttributeParamType', 'BinaryExpr', 'BooleanLiteral', 'ConfigArrayExpr', 'ConfigExpr', 'ConfigField', 'ConfigInvocationArg', 'ConfigInvocationExpr', 'DataModel', 'DataModelAttribute', 'DataModelField', 'DataModelFieldAttribute', 'DataModelFieldType', 'DataSource', 'Enum', 'EnumField', 'Expression', 'FieldInitializer', 'FunctionDecl', 'FunctionParam', 'FunctionParamType', 'GeneratorDecl', 'InternalAttribute', 'InvocationExpr', 'LiteralExpr', 'MemberAccessExpr', 'MemberAccessTarget', 'Model', 'ModelImport', 'NullExpr', 'NumberLiteral', 'ObjectExpr', 'Plugin', 'PluginField', 'ReferenceArg', 'ReferenceExpr', 'ReferenceTarget', 'StringLiteral', 'ThisExpr', 'TypeDeclaration', 'TypeDef', 'TypeDefField', 'TypeDefFieldType', 'TypeDefFieldTypes', 'UnaryExpr', 'UnsupportedFieldType'];
    }
    computeIsSubtype(subtype, supertype) {
        switch (subtype) {
            case exports.ArrayExpr:
            case exports.BinaryExpr:
            case exports.MemberAccessExpr:
            case exports.NullExpr:
            case exports.ObjectExpr:
            case exports.ReferenceExpr:
            case exports.ThisExpr:
            case exports.UnaryExpr: {
                return this.isSubtype(exports.Expression, supertype);
            }
            case exports.Attribute:
            case exports.DataSource:
            case exports.FunctionDecl:
            case exports.GeneratorDecl:
            case exports.Plugin: {
                return this.isSubtype(exports.AbstractDeclaration, supertype);
            }
            case exports.BooleanLiteral:
            case exports.NumberLiteral:
            case exports.StringLiteral: {
                return this.isSubtype(exports.LiteralExpr, supertype);
            }
            case exports.ConfigArrayExpr: {
                return this.isSubtype(exports.ConfigExpr, supertype);
            }
            case exports.DataModel: {
                return this.isSubtype(exports.AbstractDeclaration, supertype) || this.isSubtype(exports.TypeDeclaration, supertype);
            }
            case exports.DataModelField:
            case exports.TypeDefField: {
                return this.isSubtype(exports.MemberAccessTarget, supertype) || this.isSubtype(exports.ReferenceTarget, supertype);
            }
            case exports.Enum:
            case exports.TypeDef: {
                return this.isSubtype(exports.AbstractDeclaration, supertype) || this.isSubtype(exports.TypeDeclaration, supertype) || this.isSubtype(exports.TypeDefFieldTypes, supertype);
            }
            case exports.EnumField:
            case exports.FunctionParam: {
                return this.isSubtype(exports.ReferenceTarget, supertype);
            }
            case exports.InvocationExpr:
            case exports.LiteralExpr: {
                return this.isSubtype(exports.ConfigExpr, supertype) || this.isSubtype(exports.Expression, supertype);
            }
            default: {
                return false;
            }
        }
    }
    getReferenceType(refInfo) {
        const referenceId = `${refInfo.container.$type}:${refInfo.property}`;
        switch (referenceId) {
            case 'AttributeParamType:reference':
            case 'DataModelFieldType:reference':
            case 'FunctionParamType:reference': {
                return exports.TypeDeclaration;
            }
            case 'DataModel:superTypes': {
                return exports.DataModel;
            }
            case 'DataModelAttribute:decl':
            case 'DataModelFieldAttribute:decl':
            case 'InternalAttribute:decl': {
                return exports.Attribute;
            }
            case 'InvocationExpr:function': {
                return exports.FunctionDecl;
            }
            case 'MemberAccessExpr:member': {
                return exports.MemberAccessTarget;
            }
            case 'ReferenceExpr:target': {
                return exports.ReferenceTarget;
            }
            case 'TypeDefFieldType:reference': {
                return exports.TypeDefFieldTypes;
            }
            default: {
                throw new Error(`${referenceId} is not a valid reference id.`);
            }
        }
    }
    getTypeMetaData(type) {
        switch (type) {
            case 'ArrayExpr': {
                return {
                    name: 'ArrayExpr',
                    mandatory: [
                        { name: 'items', type: 'array' }
                    ]
                };
            }
            case 'Attribute': {
                return {
                    name: 'Attribute',
                    mandatory: [
                        { name: 'attributes', type: 'array' },
                        { name: 'comments', type: 'array' },
                        { name: 'params', type: 'array' }
                    ]
                };
            }
            case 'AttributeParam': {
                return {
                    name: 'AttributeParam',
                    mandatory: [
                        { name: 'attributes', type: 'array' },
                        { name: 'comments', type: 'array' },
                        { name: 'default', type: 'boolean' }
                    ]
                };
            }
            case 'AttributeParamType': {
                return {
                    name: 'AttributeParamType',
                    mandatory: [
                        { name: 'array', type: 'boolean' },
                        { name: 'optional', type: 'boolean' }
                    ]
                };
            }
            case 'ConfigArrayExpr': {
                return {
                    name: 'ConfigArrayExpr',
                    mandatory: [
                        { name: 'items', type: 'array' }
                    ]
                };
            }
            case 'ConfigInvocationExpr': {
                return {
                    name: 'ConfigInvocationExpr',
                    mandatory: [
                        { name: 'args', type: 'array' }
                    ]
                };
            }
            case 'DataModel': {
                return {
                    name: 'DataModel',
                    mandatory: [
                        { name: 'attributes', type: 'array' },
                        { name: 'comments', type: 'array' },
                        { name: 'fields', type: 'array' },
                        { name: 'isAbstract', type: 'boolean' },
                        { name: 'isView', type: 'boolean' },
                        { name: 'superTypes', type: 'array' }
                    ]
                };
            }
            case 'DataModelAttribute': {
                return {
                    name: 'DataModelAttribute',
                    mandatory: [
                        { name: 'args', type: 'array' }
                    ]
                };
            }
            case 'DataModelField': {
                return {
                    name: 'DataModelField',
                    mandatory: [
                        { name: 'attributes', type: 'array' },
                        { name: 'comments', type: 'array' }
                    ]
                };
            }
            case 'DataModelFieldAttribute': {
                return {
                    name: 'DataModelFieldAttribute',
                    mandatory: [
                        { name: 'args', type: 'array' }
                    ]
                };
            }
            case 'DataModelFieldType': {
                return {
                    name: 'DataModelFieldType',
                    mandatory: [
                        { name: 'array', type: 'boolean' },
                        { name: 'optional', type: 'boolean' }
                    ]
                };
            }
            case 'DataSource': {
                return {
                    name: 'DataSource',
                    mandatory: [
                        { name: 'fields', type: 'array' }
                    ]
                };
            }
            case 'Enum': {
                return {
                    name: 'Enum',
                    mandatory: [
                        { name: 'attributes', type: 'array' },
                        { name: 'comments', type: 'array' },
                        { name: 'fields', type: 'array' }
                    ]
                };
            }
            case 'EnumField': {
                return {
                    name: 'EnumField',
                    mandatory: [
                        { name: 'attributes', type: 'array' },
                        { name: 'comments', type: 'array' }
                    ]
                };
            }
            case 'FunctionDecl': {
                return {
                    name: 'FunctionDecl',
                    mandatory: [
                        { name: 'attributes', type: 'array' },
                        { name: 'params', type: 'array' }
                    ]
                };
            }
            case 'FunctionParam': {
                return {
                    name: 'FunctionParam',
                    mandatory: [
                        { name: 'optional', type: 'boolean' }
                    ]
                };
            }
            case 'FunctionParamType': {
                return {
                    name: 'FunctionParamType',
                    mandatory: [
                        { name: 'array', type: 'boolean' }
                    ]
                };
            }
            case 'GeneratorDecl': {
                return {
                    name: 'GeneratorDecl',
                    mandatory: [
                        { name: 'fields', type: 'array' }
                    ]
                };
            }
            case 'InternalAttribute': {
                return {
                    name: 'InternalAttribute',
                    mandatory: [
                        { name: 'args', type: 'array' }
                    ]
                };
            }
            case 'InvocationExpr': {
                return {
                    name: 'InvocationExpr',
                    mandatory: [
                        { name: 'args', type: 'array' }
                    ]
                };
            }
            case 'Model': {
                return {
                    name: 'Model',
                    mandatory: [
                        { name: 'declarations', type: 'array' },
                        { name: 'imports', type: 'array' }
                    ]
                };
            }
            case 'ObjectExpr': {
                return {
                    name: 'ObjectExpr',
                    mandatory: [
                        { name: 'fields', type: 'array' }
                    ]
                };
            }
            case 'Plugin': {
                return {
                    name: 'Plugin',
                    mandatory: [
                        { name: 'fields', type: 'array' }
                    ]
                };
            }
            case 'ReferenceExpr': {
                return {
                    name: 'ReferenceExpr',
                    mandatory: [
                        { name: 'args', type: 'array' }
                    ]
                };
            }
            case 'TypeDef': {
                return {
                    name: 'TypeDef',
                    mandatory: [
                        { name: 'attributes', type: 'array' },
                        { name: 'comments', type: 'array' },
                        { name: 'fields', type: 'array' }
                    ]
                };
            }
            case 'TypeDefField': {
                return {
                    name: 'TypeDefField',
                    mandatory: [
                        { name: 'attributes', type: 'array' },
                        { name: 'comments', type: 'array' }
                    ]
                };
            }
            case 'TypeDefFieldType': {
                return {
                    name: 'TypeDefFieldType',
                    mandatory: [
                        { name: 'array', type: 'boolean' },
                        { name: 'optional', type: 'boolean' }
                    ]
                };
            }
            default: {
                return {
                    name: type,
                    mandatory: []
                };
            }
        }
    }
}
exports.ZModelAstReflection = ZModelAstReflection;
exports.reflection = new ZModelAstReflection();
