// @zenstackhq/server/sveltekit/handler.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.default = createHandler;
exports.SvelteKitHandler = createHandler;
const rpc_1 = __importDefault(require("../api/rpc"));
const utils_1 = require("../api/utils");
const shared_1 = require("../shared");
/**
 * SvelteKit server hooks handler for handling CRUD requests.
 */
function createHandler(options) {
    (0, utils_1.logInfo)(options.logger, `ZenStackHandler installing routes at prefix: ${options.prefix}`);
    const { modelMeta, zodSchemas } = (0, shared_1.loadAssets)(options);
    const requestHandler = options.handler ?? (0, rpc_1.default)();
    return async ({ event, resolve }) => {
        if (event.url.pathname.startsWith(options.prefix)) {
            const prisma = (await options.getPrisma(event));
            if (!prisma) {
                return new Response(JSON.stringify({ message: 'unable to get prisma from request context' }), {
                    status: 400,
                    headers: {
                        'content-type': 'application/json',
                    },
                });
            }
            const query = Object.fromEntries(event.url.searchParams);
            let requestBody;
            if (event.request.body) {
                const text = await event.request.text();
                if (text) {
                    requestBody = JSON.parse(text);
                }
            }
            const path = event.url.pathname.substring(options.prefix.length);
            try {
                const r = await requestHandler({
                    method: event.request.method,
                    path,
                    query,
                    requestBody,
                    prisma,
                    modelMeta,
                    zodSchemas,
                    logger: options.logger,
                });
                return new Response(JSON.stringify(r.body), {
                    status: r.status,
                    headers: {
                        'content-type': 'application/json',
                    },
                });
            }
            catch (err) {
                return new Response(JSON.stringify({ message: `An unhandled error occurred: ${err}` }), {
                    status: 500,
                    headers: {
                        'content-type': 'application/json',
                    },
                });
            }
        }
        return resolve(event);
    };
}
