// @zenstackhq/runtime/enhancements/edge/utils.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.formatObject = formatObject;
exports.prismaClientValidationError = prismaClientValidationError;
exports.prismaClientKnownRequestError = prismaClientKnownRequestError;
exports.prismaClientUnknownRequestError = prismaClientUnknownRequestError;
exports.isUnsafeMutate = isUnsafeMutate;
exports.isAutoIncrementIdField = isAutoIncrementIdField;
const safe_json_stringify_1 = __importDefault(require("safe-json-stringify"));
const cross_1 = require("../../cross");
/**
 * Formats an object for pretty printing.
 */
// eslint-disable-next-line @typescript-eslint/no-explicit-any
function formatObject(value, multiLine = true) {
    return (0, safe_json_stringify_1.default)(value, (_, v) => (typeof v === 'bigint' ? v.toString() : v), multiLine ? 2 : undefined);
}
// eslint-disable-next-line @typescript-eslint/no-explicit-any
function prismaClientValidationError(prisma, prismaModule, message) {
    throw new prismaModule.PrismaClientValidationError(message, { clientVersion: prisma._clientVersion });
}
// eslint-disable-next-line @typescript-eslint/no-explicit-any
function prismaClientKnownRequestError(prisma, prismaModule, ...args) {
    return new prismaModule.PrismaClientKnownRequestError(...args);
}
// eslint-disable-next-line @typescript-eslint/no-explicit-any
function prismaClientUnknownRequestError(prismaModule, ...args) {
    throw new prismaModule.PrismaClientUnknownRequestError(...args);
}
// eslint-disable-next-line @typescript-eslint/no-explicit-any
function isUnsafeMutate(model, args, modelMeta) {
    if (!args) {
        return false;
    }
    for (const k of Object.keys(args)) {
        const field = (0, cross_1.resolveField)(modelMeta, model, k);
        if (field && (isAutoIncrementIdField(field) || field.isForeignKey)) {
            return true;
        }
    }
    return false;
}
function isAutoIncrementIdField(field) {
    return field.isId && field.isAutoIncrement;
}
